src/hnswutils.c

```diff
@@ -564,6 +566,179 @@ AddToVisited(HTAB *v, HnswCandidate * hc, Relation index, bool *found)
 	}
 }
 
+
+static void
+AddCandidate(IndexScanDesc scan, int lc, HnswCandidate *hc)
+{
+	HnswScanOpaque so = (HnswScanOpaque) scan->opaque;
+	LayerScanDesc* layer = &so->layers[lc];
+	bool		visited;
+	AddToVisited(layer->v, hc, scan->indexRelation, &visited);
+	if (!visited)
+	{
+		pairingheap_add(layer->C, &(CreatePairingHeapNode(hc)->ph_node));
+		pairingheap_add(layer->W, &(CreatePairingHeapNode(hc)->ph_node));
+		layer->n += lc != 0 || list_length(hc->element->heaptids) != 0;
+	}
+}
+
+void
+HnswInitScan(IndexScanDesc scan, Datum q)
+{
+	HnswScanOpaque so = (HnswScanOpaque) scan->opaque;
+	HnswElement entryPoint;
+	Relation	index = scan->indexRelation;
+	int         i;
+	HASHCTL		hash_ctl;
+	HnswCandidate *hc;
+
+	/* Create hash table */
+	hash_ctl.hcxt = CurrentMemoryContext;
+	if (index == NULL)
+	{
+		hash_ctl.keysize = sizeof(HnswElement *);
+		hash_ctl.entrysize = sizeof(HnswElement *);
+	}
+	else
+	{
+		hash_ctl.keysize = sizeof(ItemPointerData);
+		hash_ctl.entrysize = sizeof(ItemPointerData);
+	}
+
+	so->hc = NULL;
+	so->q = q;
+
+	/* Get m and entry point */
+	HnswGetMetaPageInfo(index, &so->m, &entryPoint);
+	if (entryPoint == NULL)
+	{
+		so->n_layers = 0;
+		so->layers = NULL;
+		return;
+	}
+	hc = HnswEntryCandidate(entryPoint, q, index, so->procinfo, so->collation, false);
+	so->n_layers = entryPoint->level + 1;
+	so->layers = (LayerScanDesc*)palloc(sizeof(LayerScanDesc) * so->n_layers);
+	for (i = 0; i < so->n_layers; i++)
+	{
+		so->layers[i].C = pairingheap_allocate(CompareNearestCandidates, NULL);
+		so->layers[i].W = pairingheap_allocate(CompareNearestCandidates, NULL);
```

Comment:
W should use `CompareFurthestCandidates`

Reply:
> W should use `CompareFurthestCandidates`

Actually it is not completely true. pgvector really uses reverse (distance descending) order for W, because 
1.  It needs to get largest distance already selected in W
2. Be able to replace item with largest distance with better candidate.

We do not need 2 and we need to extract best candidate from W (with smallest distance).
In the original pgvector implementation it is solved by extracting all elements and  appending them to the list (so finally in the list them are in correct distance ascending order).
But in case of iterator I need to select just one best  item.

But you was right - I incorrectly used pairingheap_first to get the largest distance. It is fixed now.

Reply:
Yep...
Using max_distance was not good idea: it cause extraction of almost all elements in memory which leads to very bad performance and high memory consumption. So I have changed implementation to use distance to n-th element as criteria to stop recursive traversal. Now this time is almost the same as with original implementation.

To do it I have implemented function `pairingheap_nth`. It is very naive implementation: it just recursively call `pairingheap_delete_first` N times and then reconstruct the tree, adding removed elements back.
Maybe it is really better to maintain one more pairing heap, but I am not sure if it is better. At least approach with `pairingheap_nth` is much simpler.

Reply:
Makes sense. I was working on a similar thing before I saw your update. See https://github.com/hlinnaka/pgvector/tree/hnsw_iterator. I took that other approach: I split W into three separate heaps: a min-heap and max-heap to maintain the top-ef elements, and a third heap to hold the rest of the elements. Yeah, `pairingheap_nth` is certainly simpler. Perhaps we should put those three heaps behind an abstraction, to make it look like one heap that provides a fast `peek nth element` operation. 

Other changes:

- GetCandidate(): It kept the search going on all levels simultaneously, but I didn't understand how it works. The original algorithm walks down to the bottom level and forgets about the upper levels after that. I changed the iterator to do the same. It now uses HnswSearchLayer() for the upper levels like usual, and only uses the "continuous mode" on the bottom level. 

- Added some comments to try to explain the continuous algorithm. Hope it helps.

What do you think?

Reply:
> I took that other approach: I split W into three separate heaps:

Yes, I also thought about it but considered that `pairingheap_nth` is more obvious approach is simpler to implement.

> GetCandidate(): It kept the search going on all levels simultaneously,

No, in my case it also performs search at one layer. It is `GetCandidate` which perform recursive traversal of all layers.

Reply:
I have tested your version of OpenAI dataset. Results are mostly the same:
```
                      _id                       |        dist        
------------------------------------------------+--------------------
 <dbpedia:Paris>                                | 0.5149979680215692
 <dbpedia:Economy_of_Paris>                     | 0.5322160352313228
 <dbpedia:List_of_tourist_attractions_in_Paris> | 0.5556925144520639
 <dbpedia:Capitale-Nationale>                   |  0.560634247353793
 <dbpedia:Besançon>                            | 0.5639481182838457
 <dbpedia:Captal>                               | 0.5744128256080486
 <dbpedia:Palais_de_Justice,_Paris>             | 0.5744524110550726
 <dbpedia:Laon>                                 | 0.5754102943098868
 <dbpedia:Hôtel_de_Ville,_Paris>               | 0.5779674934070351
 <dbpedia:Timeline_of_Paris>                    | 0.5780753294768416
(10 rows)

Time: 4.046 ms
```
A little bit slower than of original implementation (3.5 ms) but returning same top 10 documents).